pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -96,192 +143,309 @@ func shouldRestartEncryptionKeyRotation(cp *rkev1.RKEControlPlane) bool {
+				},
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationGetAPIServerStatusScript)),
+					Path:    encryptionKeyRotationGetAPIServerStatusScriptPath,
+				},
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationWaitForSecretsEncryptStatusScript)),
+					Path:    encryptionKeyRotationWaitForSecretsEncryptStatusScriptPath,
+				},
+			},
+			Instructions: []plan.OneTimeInstruction{
+				p.encryptionKeyRotationRestartApiserver(cp),
+				encryptionKeyRotationWaitForSecretsEncryptStatus(cp, ""),
+				encryptionKeyRotationSecretsEncryptStatusOneTimeInstruction(cp, ""),
+			},
+			PeriodicInstructions: []plan.PeriodicInstruction{
+				p.encryptionKeyRotationApiserverStatus(cp),
+			},
+		}
+		err := assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
+		if err != nil {
+			return p.enqueueIfErrWaiting(cp, err)
+		}
+
+		if _, ok := cp.Annotations[encryptionKeyRotationApiserverIDAnnotation]; !ok {
+			apiserverID := strings.TrimSuffix(string(leader.Plan.Output[encryptionKeyRotationRestartApiserverCommand]), "\n")
+			if len(apiserverID) != 13 { // not a valid container id
+				return p.enqueueAndSkip(cp)
+			}
+			cp.Annotations[encryptionKeyRotationApiserverIDAnnotation] = apiserverID
+			cp, err = p.rkeControlPlanes.Update(cp)
+			if err != nil {
+				return err
+			}
+			return p.enqueueAndSkip(cp)
+		}
+
+		apiserverStatus := strings.TrimSuffix(string(leader.Plan.PeriodicOutput[encryptionKeyRotationApiserverStatusCommand].Stdout), "\n")
+		if apiserverStatus == "" {
+			return p.enqueueAndSkip(cp)
+		}
+		if cp.Annotations[encryptionKeyRotationApiserverIDAnnotation] == apiserverStatus {
+			return p.enqueueAndSkip(cp)
+		}
+		if len(apiserverStatus) != 13 { // not a valid container id
+			return p.enqueueAndSkip(cp)
+		}
+	} else {
+		nodePlan := plan.NodePlan{
+			Files: []plan.File{
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationWaitForSecretsEncryptStatusScript)),
+					Path:    encryptionKeyRotationWaitForSecretsEncryptStatusScriptPath,
+				},
+			},
+			Instructions: []plan.OneTimeInstruction{
+				encryptionKeyRotationRestartInstruction(cp, ""),
+				encryptionKeyRotationWaitForSecretsEncryptStatus(cp, ""),
+				encryptionKeyRotationSecretsEncryptStatusOneTimeInstruction(cp, ""),
+			},
+		}
+		err := assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
+		if isErrWaiting(err) {
+			if strings.HasPrefix(err.Error(), "starting") {
+				logrus.Infof("rkecluster %s/%s restarting %s on %s", cp.Namespace, cp.Spec.ClusterName, rke2.GetRuntimeServerUnit(cp.Spec.KubernetesVersion), leader.Machine.Name)
+			}
+			return p.enqueueAndSkip(cp)
+		}
+	}
+
+	leaderStage, err := encryptionKeyRotationSecretsEncryptStageFromOneTimeStatus(leader)
```

Comment:
We should probably be more intelligent about this before just restarting the process otherwise we can thrash